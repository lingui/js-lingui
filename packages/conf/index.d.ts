import type { GeneratorOptions } from "@babel/core"

export declare type CatalogFormat =
  | "lingui"
  | "minimal"
  | "po"
  | "csv"
  | "po-gettext"
export type CatalogFormatOptions = {
  origins?: boolean
  lineNumbers?: boolean
}
export declare type OrderBy = "messageId" | "origin"
declare type CatalogConfig = {
  name?: string
  path: string
  include: string[]
  exclude?: string[]
}

export type LocaleObject = {
  [locale: string]: string[] | string
}

export type DefaultLocaleObject = {
  default: string
}

export declare type FallbackLocales = LocaleObject | DefaultLocaleObject

declare type CatalogService = {
  name: string
  apiKey: string
}

declare type ExtractorType = {
  match(filename: string): boolean
  extract(filename: string, targetDir: string, options?: any): void
}

export declare type LinguiConfig = {
  catalogs: CatalogConfig[]
  compileNamespace: "es" | "cjs" | "ts" | string
  extractBabelOptions: Record<string, unknown>
  compilerBabelOptions: GeneratorOptions
  fallbackLocales: FallbackLocales
  format: CatalogFormat
  extractors?: ExtractorType[]
  prevFormat: CatalogFormat
  formatOptions: CatalogFormatOptions
  localeDir: string
  locales: string[]
  catalogsMergePath: string
  orderBy: OrderBy
  pseudoLocale: string
  rootDir: string
  runtimeConfigModule:
    | [source: string, identifier?: string]
    | {
        i18n?: [source: string, identifier?: string]
        Trans?: [source: string, identifier?: string]
      }
  sourceLocale: string
  service: CatalogService
}
export declare const defaultConfig: LinguiConfig
export declare function getConfig({
  cwd,
  configPath,
  skipValidation,
}?: {
  cwd?: string
  configPath?: string
  skipValidation?: boolean
}): LinguiConfig

export declare function makeConfig(
  userConfig: Partial<LinguiConfig>,
  opts?: {
    skipValidation?: boolean
  }
): LinguiConfig

export declare const configValidation: {
  exampleConfig: {
    extractBabelOptions: {
      extends: string
      rootMode: string
      plugins: string[]
      presets: string[]
    }
    compilerBabelOptions: GeneratorOptions
    catalogs: CatalogConfig[]
    compileNamespace: "es" | "ts" | "cjs" | string
    fallbackLocales: FallbackLocales
    format: CatalogFormat
    formatOptions: CatalogFormatOptions
    locales: string[]
    orderBy: OrderBy
    pseudoLocale: string
    rootDir: string
    runtimeConfigModule: LinguiConfig["runtimeConfigModule"]
    sourceLocale: string
    service: CatalogService
  }
  deprecatedConfig: {
    fallbackLocale: (
      config: LinguiConfig & DeprecatedFallbackLanguage
    ) => string
    localeDir: (config: LinguiConfig & DeprecatedLocaleDir) => string
    srcPathDirs: (config: LinguiConfig & DeprecatedLocaleDir) => string
    srcPathIgnorePatterns: (
      config: LinguiConfig & DeprecatedLocaleDir
    ) => string
  }
  comment: string
}
export declare function replaceRootDir(
  config: LinguiConfig,
  rootDir: string
): LinguiConfig
/**
 * Replace fallbackLocale with fallbackLocales
 *
 * Released in lingui-conf 0.9
 * Remove anytime after 4.x
 */
declare type DeprecatedFallbackLanguage = {
  fallbackLocale: string | null
}

export declare function fallbackLanguageMigration(
  config: LinguiConfig & DeprecatedFallbackLanguage
): LinguiConfig
/**
 * Replace localeDir, srcPathDirs and srcPathIgnorePatterns with catalogs
 *
 * Released in @lingui/conf 3.0
 * Remove anytime after 4.x
 */
declare type DeprecatedLocaleDir = {
  localeDir: string
  srcPathDirs: string[]
  srcPathIgnorePatterns: string[]
}
export declare function catalogMigration(
  config: LinguiConfig & DeprecatedLocaleDir
): LinguiConfig
export {}
//# sourceMappingURL=index.d.ts.map
