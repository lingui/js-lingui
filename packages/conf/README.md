[![License][badge-license]][license]
[![Version][badge-version]][package]
[![Downloads][badge-downloads]][package]

# @lingui/conf

> Get lingui configuration from package.json

**⚠️ Internal package: You probably don't need this.**

`@lingui/conf` is part of [LinguiJS][linguijs]. See the [documentation][documentation] for all information, tutorials and examples.

Package finds nearest package.json starting from current directory, reads `lingui` configuration, provides defaults for all options and `<rootDir>` with current working directory.

## Usage

```js
const getConfig = require("@lingui/conf")

const config = getConfig()
// When package.json is located in '/app/package.json'
// config.rootDir === '/app'
// config.localeDir === '/app/locale'
```

See the [reference][reference] documentation of `lingui-cli` for all options.

## License

This package is licensed under [MIT][license] license.

[license]: https://github.com/lingui/js-lingui/blob/main/LICENSE
[linguijs]: https://github.com/lingui/js-lingui
[documentation]: https://lingui.dev
[reference]: https://lingui.dev/ref/conf
[package]: https://www.npmjs.com/package/@lingui/conf
[badge-downloads]: https://img.shields.io/npm/dw/@lingui/conf.svg
[badge-version]: https://img.shields.io/npm/v/@lingui/conf.svg
[badge-license]: https://img.shields.io/npm/l/@lingui/conf.svg
