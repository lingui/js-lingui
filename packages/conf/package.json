{
  "name": "@lingui/conf",
  "version": "0.0.0-managed-by-release-script",
  "description": "Get lingui configuration from package.json",
  "keywords": [
    "lingui",
    "config"
  ],
  "repository": "lingui/js-lingui",
  "bugs": "https://github.com/lingui/js-lingui/issues",
  "license": "MIT",
  "author": {
    "name": "Tomáš Ehrlich",
    "email": "tomas.ehrlich@gmail.com"
  },
  "main": "index.js",
  "engines": {
    "node": ">=4.0"
  },
  "dependencies": {
    "chalk": "^2.3.0",
    "cosmiconfig": "^4.0.0",
    "jest-regex-util": "^23.3.0",
    "jest-validate": "^23.3.0",
    "pkg-conf": "^2.1.0"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.js"
  ]
}
