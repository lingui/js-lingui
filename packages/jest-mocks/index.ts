export function getConsoleMockCalls({ mock }: jest.MockInstance<any, any>) {
  if (!mock.calls.length) return
  return mock.calls.map((call) => call[0]).join("\n")
}

export function mockConsole(
  testCase: (console: jest.Mocked<Console>) => any,
  mock = {}
) {
  function restoreConsole() {
    global.console = originalConsole
  }

  const originalConsole = global.console

  const defaults = {
    log: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
  }

  global.console = {
    ...defaults,
    ...mock,
  } as any

  let result
  try {
    result = testCase(global.console as jest.Mocked<Console>)
  } catch (e) {
    restoreConsole()
    throw e
  }

  if (result && typeof result.then === "function") {
    return result.then(restoreConsole).catch((e) => {
      restoreConsole()
      throw e
    })
  } else {
    restoreConsole()
    return result
  }
}

export function mockEnv(env, testCase) {
  const oldEnv = process.env.NODE_ENV
  process.env.NODE_ENV = env

  try {
    testCase()
  } catch (e) {
    process.env.NODE_ENV = oldEnv
    throw e
  }

  process.env.NODE_ENV = oldEnv
}
