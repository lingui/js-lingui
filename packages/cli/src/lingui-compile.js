// @flow
import chalk from "chalk"
import R from "ramda"
import program from "commander"
import plurals from "make-plural"

import { getConfig } from "@lingui/conf"

import { createCompiledCatalog } from "./api/compile"

function command(config, format, options) {
  const locales = format.getLocales()

  if (!locales.length) {
    console.log("No locales defined!\n")
    console.log(
      `(use "${chalk.yellow("lingui add-locale <language>")}" to add one)`
    )
    return false
  }

  const catalogs = R.mergeAll(
    locales.map(locale => ({ [locale]: format.read(locale) }))
  )

  const noMessages = R.compose(
    R.all(R.equals(true)),
    R.values,
    R.map(R.isEmpty)
  )
  if (noMessages(catalogs)) {
    console.log("Nothing to compile, message catalogs are empty!\n")
    console.log(
      `(use "${chalk.yellow(
        "lingui extract"
      )}" to extract messages from source files)`
    )
    return false
  }

  console.log("Compiling message catalogs…")

  return locales.map(locale => {
    const [language] = locale.split(/[_-]/)
    if (!plurals[language]) {
      console.log(
        chalk.red(
          `Error: Invalid locale ${chalk.bold(locale)} (missing plural rules)!`
        )
      )
      console.log()
      return false
    }

    const messages = R.mergeAll(
      Object.keys(catalogs[locale]).map(key => ({
        [key]: format.getTranslation(catalogs, locale, key, {
          fallbackLocale: config.fallbackLocale,
          sourceLocale: config.sourceLocale
        })
      }))
    )

    if (!options.allowEmpty && config.sourceLocale !== locale) {
      const missing = R.keys(messages).filter(
        key => messages[key] === undefined
      )

      if (missing.length) {
        console.log(
          chalk.red(
            `Error: Failed to compile catalog for locale ${chalk.bold(locale)}!`
          )
        )

        if (options.verbose) {
          console.log(chalk.red("Missing translations:"))
          missing.forEach(msgId => console.log(msgId))
        } else {
          console.log(chalk.red(`Missing ${missing.length} translation(s)`))
        }
        console.log()
        return false
      }
    }

    const compiledCatalog = createCompiledCatalog(locale, messages)
    const compiledPath = format.writeCompiled(locale, compiledCatalog)

    options.verbose && console.log(chalk.green(`${locale} ⇒ ${compiledPath}`))
    return compiledPath
  })
}

if (require.main === module) {
  program
    .description(
      "Add compile message catalogs and add language data (plurals) to compiled bundle."
    )
    .option("--strict", "Disable defaults for missing translations")
    .option("--verbose", "Verbose output")
    .option("--format <format>", "Format of message catalog")
    .on("--help", function() {
      console.log("\n  Examples:\n")
      console.log(
        "    # Compile translations and use defaults or message IDs for missing translations"
      )
      console.log("    $ lingui compile")
      console.log("")
      console.log("    # Compile translations but fail when there're missing")
      console.log("    # translations (don't replace missing translations with")
      console.log("    # default messages or message IDs)")
      console.log("    $ lingui compile --strict")
    })
    .parse(process.argv)

  const config = getConfig()
  const formatName = program.format || config.format
  const format = require(`./api/formats/${formatName}`).default(config)

  const results = command(config, format, {
    verbose: program.verbose || false,
    allowEmpty: !program.strict
  })

  if (!results || results.some(res => !res)) {
    process.exit(1)
  }

  console.log("Done!")
}
