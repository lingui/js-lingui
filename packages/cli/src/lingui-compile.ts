import chalk from "chalk"
import chokidar from "chokidar"
import fs from "fs"
import { program } from "commander"

import { getConfig, LinguiConfigNormalized } from "@lingui/conf"

import { createCompiledCatalog } from "./api/compile"
import { helpRun } from "./api/help"
import { getCatalogs, getFormat } from "./api"
import { TranslationMissingEvent } from "./api/catalog/getTranslationsForCatalog"
import { getCatalogForMerge } from "./api/catalog/getCatalogs"
import { normalizeSlashes } from "./api/utils"
import nodepath from "path"

export type CliCompileOptions = {
  verbose?: boolean
  allowEmpty?: boolean
  typescript?: boolean
  watch?: boolean
  namespace?: string
}

export async function command(
  config: LinguiConfigNormalized,
  options: CliCompileOptions
) {
  const catalogs = await getCatalogs(config)

  // Check config.compile.merge if catalogs for current locale are to be merged into a single compiled file
  const doMerge = !!config.catalogsMergePath
  let mergedCatalogs = {}

  console.log("Compiling message catalogs…")

  for (const locale of config.locales) {
    for (const catalog of catalogs) {
      const missingMessages: TranslationMissingEvent[] = []

      const messages = await catalog.getTranslations(locale, {
        fallbackLocales: config.fallbackLocales,
        sourceLocale: config.sourceLocale,
        onMissing: (missing) => {
          missingMessages.push(missing)
        },
      })

      if (!options.allowEmpty && missingMessages.length > 0) {
        console.error(
          chalk.red(
            `Error: Failed to compile catalog for locale ${chalk.bold(locale)}!`
          )
        )

        if (options.verbose) {
          console.error(chalk.red("Missing translations:"))
          missingMessages.forEach((missing) => {
            const source =
              missing.source || missing.source === missing.id
                ? `: (${missing.source})`
                : ""

            console.error(`${missing.id}${source}`)
          })
        } else {
          console.error(
            chalk.red(`Missing ${missingMessages.length} translation(s)`)
          )
        }
        console.error()
        return false
      }

      if (doMerge) {
        mergedCatalogs = { ...mergedCatalogs, ...messages }
      } else {
        const namespace = options.typescript
          ? "ts"
          : options.namespace || config.compileNamespace
        const compiledCatalog = createCompiledCatalog(locale, messages, {
          strict: false,
          namespace,
          pseudoLocale: config.pseudoLocale,
          compilerBabelOptions: config.compilerBabelOptions,
        })

        let compiledPath = await catalog.writeCompiled(
          locale,
          compiledCatalog,
          namespace
        )

        if (options.typescript) {
          const typescriptPath = compiledPath.replace(/\.ts?$/, "") + ".d.ts"
          fs.writeFileSync(
            typescriptPath,
            `import { Messages } from '@lingui/core';
          declare const messages: Messages;
          export { messages };
          `
          )
        }

        compiledPath = normalizeSlashes(
          nodepath.relative(config.rootDir, compiledPath)
        )

        options.verbose &&
          console.error(chalk.green(`${locale} ⇒ ${compiledPath}`))
      }
    }

    if (doMerge) {
      const compileCatalog = await getCatalogForMerge(config)
      const namespace = options.namespace || config.compileNamespace
      const compiledCatalog = createCompiledCatalog(locale, mergedCatalogs, {
        strict: false,
        namespace: namespace,
        pseudoLocale: config.pseudoLocale,
        compilerBabelOptions: config.compilerBabelOptions,
      })
      let compiledPath = await compileCatalog.writeCompiled(
        locale,
        compiledCatalog,
        namespace
      )

      compiledPath = normalizeSlashes(
        nodepath.relative(config.rootDir, compiledPath)
      )

      options.verbose && console.log(chalk.green(`${locale} ⇒ ${compiledPath}`))
    }
  }
  return true
}

type CliOptions = {
  verbose?: boolean
  allowEmpty?: boolean
  typescript?: boolean
  watch?: boolean
  namespace?: string
  strict?: string
  config?: string
  debounce?: number
}

if (require.main === module) {
  program
    .description(
      "Add compile message catalogs and add language data (plurals) to compiled bundle."
    )
    .option("--config <path>", "Path to the config file")
    .option("--strict", "Disable defaults for missing translations")
    .option("--verbose", "Verbose output")
    .option("--typescript", "Create Typescript definition for compiled bundle")
    .option(
      "--namespace <namespace>",
      "Specify namespace for compiled bundle. Ex: cjs(default) -> module.exports, es -> export, window.test -> window.test"
    )
    .option("--watch", "Enables Watch Mode")
    .option(
      "--debounce <delay>",
      "Debounces compilation for given amount of milliseconds"
    )
    .on("--help", function () {
      console.log("\n  Examples:\n")
      console.log(
        "    # Compile translations and use defaults or message IDs for missing translations"
      )
      console.log(`    $ ${helpRun("compile")}`)
      console.log("")
      console.log("    # Compile translations but fail when there are missing")
      console.log("    # translations (don't replace missing translations with")
      console.log("    # default messages or message IDs)")
      console.log(`    $ ${helpRun("compile --strict")}`)
    })
    .parse(process.argv)

  const options = program.opts<CliOptions>()

  const config = getConfig({ configPath: options.config })

  let previousRun = Promise.resolve(true)

  const compile = () => {
    previousRun = previousRun.then(() =>
      command(config, {
        verbose: options.watch || options.verbose || false,
        allowEmpty: !options.strict,
        typescript:
          options.typescript || config.compileNamespace === "ts" || false,
        namespace: options.namespace, // we want this to be undefined if user does not specify so default can be used
      })
    )

    return previousRun
  }

  let debounceTimer: NodeJS.Timer

  const dispatchCompile = () => {
    // Skip debouncing if not enabled
    if (!options.debounce) compile()

    // CLear the previous timer if there is any, and schedule the next
    debounceTimer && clearTimeout(debounceTimer)
    debounceTimer = setTimeout(() => compile(), options.debounce)
  }

  // Check if Watch Mode is enabled
  if (options.watch) {
    console.info(chalk.bold("Initializing Watch Mode..."))
    ;(async function initWatch() {
      const format = await getFormat(
        config.format,
        config.formatOptions,
        config.sourceLocale
      )
      const catalogs = await getCatalogs(config)

      const paths: string[] = []

      config.locales.forEach((locale) => {
        catalogs.forEach((catalog) => {
          paths.push(
            `${catalog.path
              .replace(/{locale}/g, locale)
              .replace(/{name}/g, "*")}${format.getCatalogExtension()}`
          )
        })
      })

      const watcher = chokidar.watch(paths, {
        persistent: true,
      })

      const onReady = () => {
        console.info(chalk.green.bold("Watcher is ready!"))
        watcher
          .on("add", () => dispatchCompile())
          .on("change", () => dispatchCompile())
      }

      watcher.on("ready", () => onReady())
    })()
  } else {
    compile().then((results) => {
      if (!results) {
        process.exit(1)
      }

      console.log("Done!")
    })
  }
}
