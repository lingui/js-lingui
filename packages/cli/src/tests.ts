import os from "os"
import path from "path"
import fs from "fs"

import {
  Catalog,
  MakeOptions,
  MakeTemplateOptions,
  MergeOptions,
} from "./api/catalog"
import { LinguiConfig, makeConfig } from "@lingui/conf"
import { ExtractedMessageType, getFormat, MessageType } from "./api"

export async function copyFixture(fixtureDir: string) {
  const tmpDir = await fs.promises.mkdtemp(
    path.join(os.tmpdir(), `lingui-test-${process.pid}`)
  )

  try {
    await fs.promises.cp(fixtureDir, tmpDir, { recursive: true })
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code != "ENOENT") {
      throw err
    }
  }
  return tmpDir
}

export const defaultMakeOptions: MakeOptions = {
  verbose: false,
  clean: false,
  overwrite: false,
  locale: null,
  prevFormat: null,
  orderBy: "messageId",
}

export const defaultMakeTemplateOptions: MakeTemplateOptions = {
  verbose: false,
  orderBy: "messageId",
}

export const defaultMergeOptions: MergeOptions = {
  overwrite: false,
}

// on windows line endings are different,
// so direct comparison to snapshot would file if not normalized
export const normalizeLineEndings = (str: string) =>
  str.replace(/\r?\n/g, "\r\n")

export const makeCatalog = async (_config: Partial<LinguiConfig> = {}) => {
  const config = makeConfig(_config, { skipValidation: true })

  return new Catalog(
    {
      name: "messages",
      path: "{locale}/messages",
      include: [],
      exclude: [],
      format: await getFormat(
        config.format,
        config.formatOptions,
        config.sourceLocale
      ),
    },
    config
  )
}

export function makePrevMessage(message = {}): MessageType {
  return {
    translation: "",
    ...makeNextMessage(message),
  }
}

export function makeNextMessage(message = {}): ExtractedMessageType {
  return {
    origin: [["catalog.test.ts", 1]],
    obsolete: false,
    ...message,
  }
}

type Listing = { [filename: string]: string | Listing }

export function listingToHumanReadable(listing: Listing): string {
  const output: string[] = []
  Object.entries(listing).forEach(([filename, value]) => {
    if (typeof value === "string") {
      output.push("#######################")
      output.push(`Filename: ${filename}`)
      output.push("#######################")
      output.push("")
      output.push(normalizeLineEndings(value))
      output.push("")
    } else {
      output.push(...listingToHumanReadable(value))
    }
  })

  return output.join("\n")
}

/**
 * Create fixtures from provided listing in temp folder
 * Alternative for mock-fs which is also mocking nodejs require calls
 * @param listing
 */
export async function createFixtures(listing: Listing) {
  const tmpDir = await fs.promises.mkdtemp(
    path.join(os.tmpdir(), `lingui-test-${process.pid}`)
  )

  async function create(listing: Listing) {
    for (const [filename, value] of Object.entries(listing)) {
      if (typeof value === "string") {
        await fs.promises.writeFile(path.join(tmpDir, filename), value)
      } else {
        await create(value)
      }
    }
  }

  await create(listing)
  return tmpDir
}

export function readFsToJson(
  directory: string,
  filter?: (filename: string) => boolean
): Listing {
  const out: Listing = {}

  fs.readdirSync(directory).map((filename) => {
    const filepath = path.join(directory, filename)

    if (fs.lstatSync(filepath).isDirectory()) {
      out[filename] = readFsToJson(filepath)
      return out
    }

    if (!filter || filter(filename)) {
      out[filename] = fs.readFileSync(filepath).toString()
    }
  })

  return out
}
