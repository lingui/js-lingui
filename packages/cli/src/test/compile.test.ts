import { command } from "../lingui-compile"
import { makeConfig } from "@lingui/conf"
import { getConsoleMockCalls, mockConsole } from "@lingui/jest-mocks"
import { createFixtures, readFsToJson } from "../tests"

describe("CLI Command: Compile", () => {
  describe("Merge Catalogs", () => {
    // todo
  })

  function getConfig(rootDir: string) {
    return makeConfig({
      locales: ["en", "pl"],
      sourceLocale: "en",
      rootDir: rootDir,
      catalogs: [
        {
          path: "<rootDir>/{locale}",
          include: ["<rootDir>"],
          exclude: [],
        },
      ],
    })
  }

  describe("allowEmpty = false", () => {
    it(
      "Should show error and stop compilation of catalog " +
        "if message doesnt have a translation (no template)",
      async () => {
        expect.assertions(4)

        const rootDir = await createFixtures({
          "/test": {
            "en.po": `
msgid "Hello World"
msgstr "Hello World"
        `,
            "pl.po": `
msgid "Hello World"
msgstr "Cześć świat"

msgid "Test String"
msgstr ""
        `,
          },
        })

        const config = getConfig(rootDir)

        await mockConsole(async (console) => {
          const result = await command(config, {
            allowEmpty: false,
          })
          const actualFiles = readFsToJson(config.rootDir)

          expect(actualFiles["pl.js"]).toBeFalsy()
          expect(actualFiles["en.js"]).toBeTruthy()

          const log = getConsoleMockCalls(console.error)
          expect(log).toMatchSnapshot()
          expect(result).toBeFalsy()
        })
      }
    )

    it(
      "Should show error and stop compilation of catalog " +
        "if message doesnt have a translation (with template)",
      async () => {
        expect.assertions(3)
        const rootDir = await createFixtures({
          "/test": {
            "messages.pot": `
msgid "Hello World"
msgstr ""
        `,
            "pl.po": ``,
          },
        })

        const config = getConfig(rootDir)

        await mockConsole(async (console) => {
          const result = await command(config, {
            allowEmpty: false,
          })

          const actualFiles = readFsToJson(rootDir)

          expect({
            pl: actualFiles["pl.js"],
            en: actualFiles["en.js"],
          }).toMatchSnapshot()

          const log = getConsoleMockCalls(console.error)
          expect(log).toMatchSnapshot()
          expect(result).toBeFalsy()
        })
      }
    )

    it("Should show missing messages verbosely when verbose = true", async () => {
      expect.assertions(2)
      const rootDir = await createFixtures({
        "/test": {
          "pl.po": `
msgid "Hello World"
msgstr ""

msgid "Test String"
msgstr ""
        `,
        },
      })

      const config = getConfig(rootDir)

      await mockConsole(async (console) => {
        const result = await command(config, {
          allowEmpty: false,
          verbose: true,
        })

        const log = getConsoleMockCalls(console.error)
        expect(log).toMatchSnapshot()
        expect(result).toBeFalsy()
      })
    })
  })
})
