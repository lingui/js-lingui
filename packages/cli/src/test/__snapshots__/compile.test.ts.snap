// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI Command: Compile allowEmpty = false Should show error and stop compilation of catalog if message doesnt have a translation (no template) 1`] = `
Error: Failed to compile catalog for locale pl!
Missing 1 translation(s)

`;

exports[`CLI Command: Compile allowEmpty = false Should show error and stop compilation of catalog if message doesnt have a translation (with template) 1`] = `
{
  en: undefined,
  pl: undefined,
}
`;

exports[`CLI Command: Compile allowEmpty = false Should show error and stop compilation of catalog if message doesnt have a translation (with template) 2`] = `
Error: Failed to compile catalog for locale en!
Missing 1 translation(s)

`;

exports[`CLI Command: Compile allowEmpty = false Should show missing messages verbosely when verbose = true 1`] = `
en â‡’ en.js
Error: Failed to compile catalog for locale pl!
Missing translations:
mY42CM: (Hello World)
2ZeN02: (Test String)

`;
